README.md

```diff
@@ -111,6 +124,268 @@ instructions in its
 (coordinating guidance is in comments in the
 default recipe).
 
+
+## Other Guidelines
+
+### Modules and classes in cookbooks
+
+It is often useful to factor out logic into a library - especially logic that
+doesn't create resources. Doing so makes this logic easier to unit test and
+makes the recipe or resource cleaner.
+
+Our standard is that all cookbooks use the top-level container of `module FB`,
+and then create a class for their cookbook under that. For example, `fb_fstab`
+creates a `class Fstab` inside of the `module FB`. We will refer to this as
+the cookbook class from here.
+
+We require all cookbooks use this model for consitency.
+
+Since we don't put anything other than other classes inside the top-level
+object, it's clear that a `module` is the right choice.
+
+While there is no reason that a cookbook class can't be one design to be
+instantiated, more often than not it is simply a collection of callable class
+methods - a place to put methods that can then be called both from this
+cookbook and others.
+
+Below the cookbook class, the author is free to make whatever class or methods
+they desire.
+
+It is often desirable to `include` a module into a Custom Resource, and we do
+this by creating a module inside of the cookbook class, often ending in
+`Provider`, e.g. `FstabProvider`. This code should be in its own library file.
+
+It is worth noting that `fb_fstab` itself violates this guideline as it is
+quite old before we had solidified a standard and thus it's provider module
+is directly below the `module FB`.
+
+### Extending the node vs self-contained classes
+
+You may have noticed that some of our cookbooks will extend the `node` object,
+while others have self-contained classes that sometimes require the `node` be
+passed as a parameter to some methods.
+
+In general, the **only** time when extending the `node` is acceptable is when
+you are simply making a convenience function around using the node object. So,
```

Comment:
I agree with this, and we both know what you mean here, but "convenience function" isn't much of a guideline for someone that's not already familiar with this stuff. Can/should they consume ohai data? node data from the cookbook? from other cookbooks? Can they do if then else logic or should it just be one line? Note that we currently have all of the above in the codebase.

Reply:
This is why I included some examples. I think it's impossible to provide a hard line, this is mostly going to be up to our judgement on PR reviews, but this is a generally useful starting point, I think.